## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [31]:
import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np

In [32]:
import seaborn as sns


titanic_data = sns.load_dataset("titanic")
titanic_data.tail(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
886,0,2,male,27.0,0,0,13.00,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.00,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [33]:
miss = titanic_data.isnull().sum()
print(miss)

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [34]:
titanic_data = titanic_data.dropna(thresh=len(titanic_data) / 2, axis='columns')
titanic_data = titanic_data.dropna(thresh=len(titanic_data.columns) / 2, axis='index')
titanic_data

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [35]:
mask_man = titanic_data['who'] == 'man'
mask_woman = titanic_data['who'] == 'woman'
mask_child = titanic_data['who'] == 'child'

median_man = round(titanic_data[mask_man]['age'].median())
median_woman = round(titanic_data[mask_woman]['age'].median())
median_child = round(titanic_data[mask_child]['age'].median())

titanic_data.loc[(mask_man) & (titanic_data['age'].isnull()), 'age'] = median_man
titanic_data.loc[(mask_woman) & (titanic_data['age'].isnull()), 'age'] = median_woman
titanic_data.loc[(mask_child) & (titanic_data['age'].isnull()), 'age'] = median_child
print(titanic_data)

     survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female  30.0      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  


### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [36]:
titanic_data.dropna(thresh=len(titanic_data.columns) - 1)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,30.0,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [37]:
town = titanic_data['embark_town'].value_counts().idxmax()

print(town)

Southampton


### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [38]:
count_of_survived = titanic_data['survived'].sum()
count_ = len(titanic_data.index)

percent = round((count_of_survived / count_) * 100, 2)
print(percent)

38.38


### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [39]:
count_srv_in_town = titanic_data[titanic_data['survived'] == 1]['embark_town'].value_counts()
print(count_srv_in_town)

embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: count, dtype: int64


### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [40]:
count_srv_class = titanic_data[titanic_data['survived'] == 1]['class'].value_counts()

_class = titanic_data['class'].value_counts()

percent_srv_class = ((count_srv_class / _class) * 100).round(2)
print(percent_srv_class)

class
First     62.96
Second    47.28
Third     24.24
Name: count, dtype: float64


### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [23]:
rich_mask = titanic_data['fare'] >= 100
count_rich = len(titanic_data[rich_mask])

count_rich_srv = len(titanic_data[rich_mask & (titanic_data['survived'] == 1)])

rich_survival_precent = round((count_rich_srv / count_rich) * 100, 2)
print(rich_survival_precent)

73.58


### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [41]:
mask_child = titanic_data['who'] == 'child'

count_child_alon = len(titanic_data[mask_child & titanic_data['alone']]['who'])

Какие выводы вы можете сделать о выживших пассажирах Титаника? 

In [ ]:
figure, axis = plt.subplots(1, 2, figsize=(19, 12))
axis: np.ndarray

wed, _, _ = axis[1].pie(
    count_srv_in_town,
    autopct='%1.1f%%', 
    colors=['#ff9999',"#8066ff",'#99ff99'],
    shadow = True
)

labels = count_srv_in_town.index
axis[1].set_title("Выжившие по городам", fontsize=18, fontweight="bold")

axis[1].legend(wed, labels, title="Towns", loc="center left", bbox_to_anchor=(1, 0.5, 0.25, 0.75))

bar = axis[0].bar(
    np.arange(percent_srv_class.size),
    percent_srv_class,
    color="#5477f4"
)

axis[0].set_ylim(0, 80) 

axis[0].set_title("Выжившие по классам", fontsize=18, fontweight="bold", c="dimgray")
axis[0].set_ylabel("precent of survivors", fontsize=15, fontweight="bold", c="dimgray")
axis[0].set_xticks(
    np.arange(3),
    labels = ['First', 'Second', 'Third '],
    weight="bold"
)

plt.savefig("Titanic_conclusions.png")
